model-server/src/main/kotlin/org/modelix/model/server/handlers/ModelReplicationServer.kt

```diff
@@ -372,17 +381,34 @@
         respond(delta)
     }
 
-    private suspend fun ApplicationCall.respondDeltaAsObjectStream(versionHash: String, baseVersionHash: String?, plainText: Boolean) {
-        respondTextWriter(contentType = if (plainText) ContentType.Text.Plain else VersionDeltaStream.CONTENT_TYPE) {
-            repositoriesManager.computeDelta(versionHash, baseVersionHash).asFlow()
-                .flatten()
-                .withSeparator("\n")
-                .onEmpty { emit(versionHash) }
-                .withIndex()
-                .collect {
-                    if (it.index == 0) check(it.value == versionHash) { "First object should be the version" }
-                    append(it.value)
-                }
+    private suspend fun ApplicationCall.respondDeltaAsObjectStream(
+        versionHash: String,
+        baseVersionHash: String?,
+        plainText: Boolean,
+    ) {
+        // Call `computeDelta` before starting to respond.
+        // It could already throw an exception, and in that case we do not want a successful response status.
+        val objectData = repositoriesManager.computeDelta(versionHash, baseVersionHash)
+        respondBytesWriter(contentType = if (plainText) ContentType.Text.Plain else VersionDeltaStream.CONTENT_TYPE) {
+            try {
+                objectData.asFlow()
+                    .flatten()
+                    .withSeparator("\n")
+                    .onEmpty { emit(versionHash) }
+                    .withIndex()
+                    .collect {
+                        if (it.index == 0) check(it.value == versionHash) { "First object should be the version" }
+                        writeStringUtf8(it.value)
+                    }
+            } catch (cause: Throwable) {
```

Comment:
Any chance to be more specific here? Or at least resort to Exception and not Throwable?

Reply:
I now implemented a solution that does not need a catch block and is reusabel.

---

It seems `Throwable`s are usually caught in libary code.

> It would be more typical in a framework type application (for example an application server or a testing framework) where it can be running unknown code and should not be affected by anything that goes wrong with that code, as much as possible

See https://stackoverflow.com/questions/2274102/difference-between-using-throwable-and-exception-in-a-try-catch

> ```kt
> public inline fun ByteWriteChannel.use(block: ByteWriteChannel.() -> Unit) {
>    contract {
>        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
>    }
>
>    try {
>        block()
>    } catch (cause: Throwable) {
>        close(cause)
>        throw cause
>    } finally {
>        close()
>    }
>}
>```


See `ByteWriteChannel.use(block: ByteWriteChannel.() -> Unit)`